src/Lib.P2P/SecureCommunication/Secio1.cs

```diff
@@ -0,0 +1,246 @@
+
+#endregion
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Common.Logging;
+using Lib.P2P.Cryptography;
+using Lib.P2P.Protocols;
+using MultiFormats;
+using Org.BouncyCastle.Security;
+using ProtoBuf;
+using Semver;
+
+namespace Lib.P2P.SecureCommunication
+{
+    /// <summary>
+    ///   Creates a secure connection with a peer.
+    /// </summary>
+    public class Secio1 : IEncryptionProtocol
+    {
+        private static ILog _log = LogManager.GetLogger(typeof(Secio1));
+
+        /// <inheritdoc />
+        public string Name { get; } = "secio";
+
+        /// <inheritdoc />
+        public SemVersion Version { get; } = new SemVersion(1);
+
+        /// <inheritdoc />
+        public override string ToString() { return $"/{Name}/{Version}"; }
+
+        /// <inheritdoc />
+        public async Task ProcessMessageAsync(PeerConnection connection,
+            Stream stream,
+            CancellationToken cancel = default)
+        {
+            await EncryptAsync(connection, cancel).ConfigureAwait(false);
+        }
+
+#pragma warning disable VSTHRD103 
+        /// <inheritdoc />
+        public async Task<Stream> EncryptAsync(PeerConnection connection,
+            CancellationToken cancel = default)
+        {
+            var stream = connection.Stream;
+            var localPeer = connection.LocalPeer;
+            connection.RemotePeer = connection.RemotePeer ?? new Peer();
+            var remotePeer = connection.RemotePeer;
+
+            // =============================================================================
+            // step 1. Propose -- propose cipher suite + send pubkey + nonce
+            var rng = new SecureRandom();
+            var localNonce = new byte[16];
+            rng.NextBytes(localNonce);
+            var localProposal = new Secio1Propose
+            {
+                Nonce = localNonce,
+                Exchanges = "P-256,P-384,P-521",
+                Ciphers = "AES-256,AES-128",
+                Hashes = "SHA256,SHA512",
+                PublicKey = Convert.FromBase64String(localPeer.PublicKey)
+            };
+
+            Serializer.SerializeWithLengthPrefix(stream, localProposal, PrefixStyle.Fixed32BigEndian);
+            await stream.FlushAsync(cancel).ConfigureAwait(false);
+
+            // =============================================================================
+            // step 1.1 Identify -- get identity from their key
+            var remoteProposal =
+                Serializer.DeserializeWithLengthPrefix<Secio1Propose>(stream, PrefixStyle.Fixed32BigEndian);
+            var ridAlg = remoteProposal.PublicKey.Length <= 48 ? "identity" : "sha2-256";
+            var remoteId = MultiHash.ComputeHash(remoteProposal.PublicKey, ridAlg);
+            if (remotePeer.Id == null)
+            {
+                remotePeer.Id = remoteId;
+            }
+            
+            else if (remoteId != remotePeer.Id)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add the missing 'else' clause.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)